lib/stringtheory/src/lib.rs

```diff
@@ -261,78 +650,162 @@ impl fmt::Display for MetaString {
     }
 }
 
-/// A string that can be inlined up to 23 bytes.
-#[derive(Clone)]
-struct InlinedString {
-    data: [u8; 24],
-}
+#[cfg(test)]
+mod tests {
+    use std::num::NonZeroUsize;
 
-impl InlinedString {
-    fn new(s: &str) -> Option<Self> {
-        let s_len = s.len();
-        if s_len > 23 {
-            return None;
-        }
+    use proptest::{prelude::*, proptest};
 
-        // SAFETY: We know it fits because we just checked that the string length is 23 or less.
-        let s_len_b = s_len as u8;
+    use super::{interning::FixedSizeInterner, InlinedUnion, Inner, MetaString, UnionType};
 
-        let mut data = [0; 24];
-        data[0] = s_len_b;
+    #[test]
+    fn struct_sizes() {
+        // Overall, we expect `MetaString`, and thus `Inner`, to always be three machine words.
+        assert_eq!(std::mem::size_of::<MetaString>(), std::mem::size_of::<usize>() * 3);
+        assert_eq!(std::mem::size_of::<Inner>(), std::mem::size_of::<usize>() * 3);
+
+        // We also expect all of inlined union variant to be the exact size of `Inner`, which means we're properly
+        // maximizing the available space for inlining.
+        assert_eq!(std::mem::size_of::<InlinedUnion>(), std::mem::size_of::<Inner>());
+    }
 
-        let s_buf = s.as_bytes();
-        data[1..s_len + 1].copy_from_slice(s_buf);
+    #[test]
+    fn static_str_inlineable() {
+        let s = "hello";
+        let meta = MetaString::from_static(s);
 
-        Some(Self { data })
+        assert_eq!(s, &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Inlined);
     }
-}
 
-impl Deref for InlinedString {
-    type Target = str;
+    #[test]
+    fn static_str_not_inlineable() {
+        let s = "hello there, world! it's me, margaret!";
+        let meta = MetaString::from_static(s);
 
-    fn deref(&self) -> &str {
-        let len = self.data[0] as usize;
-        let s = &self.data[1..len + 1];
+        assert_eq!(s, &*meta);
```

Comment:
```suggestion
        assert_eq!(s, &meta);
```
Verify AutoDeref?

Reply:
It doesn't seem like autoderef kicks in here, since it's just focusing on if both sides have a lined up `PartialEq<T>` implementation.